### ISSUE TITLE
[BUG] - FeeTooSmallUTxO when registering the pool 

### ISSUE THREAD

**Issue by Do-do-new** [input-output-hk/cardano-node#2335]
**Internal/External**
*External* 

**Summary**
I am getting FeeTooSmallUTxO error when submitting transactions using the tx_fee provided by cardano-cli  transaction calculate-min-fee when registering the pool.
Normal transactions work fine.

**Steps to reproduce**
It is perfectly shown in this tutorial https://youtu.be/OHj0eJ8p4OY?t=1867
Getting the same (around 31:30 he is getting this error)

**Expected behavior**
Calculate-min-fee gives the correct fee. Otherwise we need to try to submit, extract fee and resubmit

**System info (please complete the following information):**
- OS: Ubuntu
- Version 18.04
- Node version 1.25.1
- Testnet, Mary era

**Comment by hootan-nikbakht** [comment]
What number are you using for '--witness-count' when running 'calculate-min-fee' cmdline ? 
Your '--witness-count' should be as many keys you are using during 'transaction sign' cmdline. For registering your stakepool you need 3 keys payment.skey, stake.skey, cold.skey. 

If you are indeed using 3 keys then change '--witness-count 1' change it to '--witness-count 3', this will increase the transaction fee and should resolve 'FeeTooSmallUTxO error'.

**Comment by Do-do-new** [comment]
@hootan-nikbakht Thanks for a prompt reply, I will try it out

**Comment by tigrpoolcom** [comment]
@Do-do-new : You are right its not working, even if you set all valid parameters.
What bugs me the most: if you create a draft.tx and calculate the min-fee, you will receive different results depending on the fee.
If the fee calculation is dependant on knowing the fee, you cant calulate it. Thats nonsense.

In my case i followed this [tutorial ](https://docs.cardano.org/projects/cardano-node/en/latest/stake-pool-operations/register_key.html) for registering a stake address:

Try one: --invalid-hereafter 0 --fee 0 then submit
```
dd2d13c29f7e84ef4bd97f2dde7e412e79556687ad520beb0c1d1278a6100933 0 5000000 lovelace
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli transaction build-raw --tx-in dd2d13c29f7e84ef4bd97f2dde7e412e79556687ad520beb0c1d1278a6100933#0 --tx-out $(cat payment.addr)+0 --invalid-hereafter 0 --fee 0 --out-file tx.draft --certificate-file stake.cert;"
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli query protocol-parameters --mary-era --$CDN_NETCMD --out-file protocol.json;"
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli transaction calculate-min-fee --tx-body-file tx.draft --tx-in-count 1 --tx-out-count 1 --witness-count 1 --protocol-params-file protocol.json;"
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Output Result: 172453 Lovelace
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli query protocol-parameters --mary-era --$CDN_NETCMD --out-file protocol.json;"
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "awk '/keyDeposit/{print $2}' ${CDN_DIR}/protocol.json "
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Output Result: 2000000,
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli query tip --${CDN_NETCMD} | awk '/slotNo/{print $2}' "
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Output Result: 19029200
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli transaction build-raw --tx-in dd2d13c29f7e84ef4bd97f2dde7e412e79556687ad520beb0c1d1278a6100933#0 --tx-out $(cat payment.addr)+2827547 --invalid-hereafter 19029400 --fee 172453 --out-file tx.raw --certificate-file stake.cert;"
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli transaction sign --tx-body-file tx.raw --signing-key-file payment.skey --signing-key-file stake.skey --signing-key-file cold.skey --out-file tx.signed;"
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli transaction submit --tx-file tx.signed --$CDN_NETCMD ;"
fail: CardanoDirector.BusinessLayer.CommandExecutor[0]
Error Result: Shelley command failed: transaction submit Error: Error while submitting tx: ApplyTxError [LedgerFailure (UtxowFailure (UtxoFailure (OutsideValidityIntervalUTxO (ValidityInterval {invalidBefore = SNothing, invalidHereafter = SJust (SlotNo 19029400)}) (SlotNo 19030460)))),LedgerFailure (UtxowFailure (UtxoFailure (FeeTooSmallUTxO (Coin 175709) (Coin 172453))))]
```

Try two:  --invalid-hereafter 19033000 --fee 0 then submit
```
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli stake-address registration-certificate --stake-verification-key-file stake.vkey --out-file stake.cert;"
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli query utxo --mary-era --address $(cat payment.addr) --${CDN_NETCMD} ;"
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Output Result: TxHash TxIx Amount
--------------------------------------------------------------------------------------
dd2d13c29f7e84ef4bd97f2dde7e412e79556687ad520beb0c1d1278a6100933 0 5000000 lovelace

info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli query tip --${CDN_NETCMD} | awk '/slotNo/{print $2}' "
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Output Result: 19032800

info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli transaction build-raw --tx-in dd2d13c29f7e84ef4bd97f2dde7e412e79556687ad520beb0c1d1278a6100933#0 --tx-out $(cat payment.addr)+0 --invalid-hereafter 19033000 --fee 0 --out-file tx.draft --certificate-file stake.cert;"
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli query protocol-parameters --mary-era --$CDN_NETCMD --out-file protocol.json;"
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli transaction calculate-min-fee --tx-body-file tx.draft --tx-in-count 1 --tx-out-count 1 --witness-count 1 --protocol-params-file protocol.json;"
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Output Result: 172629 Lovelace

info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli query protocol-parameters --mary-era --$CDN_NETCMD --out-file protocol.json;"
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "awk '/keyDeposit/{print $2}' ${CDN_DIR}/protocol.json "
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Output Result: 2000000,

info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli query tip --${CDN_NETCMD} | awk '/slotNo/{print $2}' "
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Output Result: 19032800

info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli transaction build-raw --tx-in dd2d13c29f7e84ef4bd97f2dde7e412e79556687ad520beb0c1d1278a6100933#0 --tx-out $(cat payment.addr)+2827371 --invalid-hereafter 19033000 --fee 172629 --out-file tx.raw --certificate-file stake.cert;"
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli transaction sign --tx-body-file tx.raw --signing-key-file payment.skey --signing-key-file stake.skey --signing-key-file cold.skey --out-file tx.signed;"
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli transaction submit --tx-file tx.signed --$CDN_NETCMD ;"
fail: CardanoDirector.BusinessLayer.CommandExecutor[0]
Error Result: Shelley command failed: transaction submit Error: Error while submitting tx: ApplyTxError [LedgerFailure (UtxowFailure (UtxoFailure (FeeTooSmallUTxO (Coin 175709) (Coin 172629))))]
```

Try three: --invalid-hereafter 19035185 --fee 0 then --invalid-hereafter 19035185 --fee 172805 then submit
```
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli stake-address registration-certificate --stake-verification-key-file stake.vkey --out-file stake.cert;"
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli query utxo --mary-era --address $(cat payment.addr) --${CDN_NETCMD} ;"
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Output Result: TxHash TxIx Amount
--------------------------------------------------------------------------------------
dd2d13c29f7e84ef4bd97f2dde7e412e79556687ad520beb0c1d1278a6100933 0 5000000 lovelace
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli query tip --${CDN_NETCMD} | awk '/slotNo/{print $2}' "
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Output Result: 19034985
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli transaction build-raw --tx-in dd2d13c29f7e84ef4bd97f2dde7e412e79556687ad520beb0c1d1278a6100933#0 --tx-out $(cat payment.addr)+5000000 --invalid-hereafter 19035185 --fee 0 --out-file tx.draft --certificate-file stake.cert;"
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli query protocol-parameters --mary-era --$CDN_NETCMD --out-file protocol.json;"
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli transaction calculate-min-fee --tx-body-file tx.draft --tx-in-count 1 --tx-out-count 1 --witness-count 1 --protocol-params-file protocol.json;"
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Output Result: 172805 Lovelace
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli query tip --${CDN_NETCMD} | awk '/slotNo/{print $2}' "
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Output Result: 19034985
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli transaction build-raw --tx-in dd2d13c29f7e84ef4bd97f2dde7e412e79556687ad520beb0c1d1278a6100933#0 --tx-out $(cat payment.addr)+5000000 --invalid-hereafter 19035185 --fee 172805 --out-file tx.draft --certificate-file stake.cert;"
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli query protocol-parameters --mary-era --$CDN_NETCMD --out-file protocol.json;"
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli transaction calculate-min-fee --tx-body-file tx.draft --tx-in-count 1 --tx-out-count 1 --witness-count 1 --protocol-params-file protocol.json;"
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Output Result: 172981 Lovelace
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli query protocol-parameters --mary-era --$CDN_NETCMD --out-file protocol.json;"
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "awk '/keyDeposit/{print $2}' ${CDN_DIR}/protocol.json "
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Output Result: 2000000,
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli query tip --${CDN_NETCMD} | awk '/slotNo/{print $2}' "
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Output Result: 19034985
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli transaction build-raw --tx-in dd2d13c29f7e84ef4bd97f2dde7e412e79556687ad520beb0c1d1278a6100933#0 --tx-out $(cat payment.addr)+2827019 --invalid-hereafter 19035185 --fee 172981 --out-file tx.raw --certificate-file stake.cert;"
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli transaction sign --tx-body-file tx.raw --signing-key-file payment.skey --signing-key-file stake.skey --signing-key-file cold.skey --out-file tx.signed;"
info: CardanoDirector.BusinessLayer.CommandExecutor[0]
Executed Command: /bin/bash -c "cardano-cli transaction submit --tx-file tx.signed --$CDN_NETCMD ;"
fail: CardanoDirector.BusinessLayer.CommandExecutor[0]
Error Result: Shelley command failed: transaction submit Error: Error while submitting tx: ApplyTxError [LedgerFailure (UtxowFailure (UtxoFailure (FeeTooSmallUTxO (Coin 175709) (Coin 172981))))]
```

**Comment by tigrpoolcom** [comment]
@hootan-nikbakht : I saw that in my signing i had three certificates and witness was only 1.

I've changed that and now were able to submit the transaction.
BUT: It looks like the fee calculation is more a guess then a real calulation.

```
Executed Command: /bin/bash -c "cardano-cli stake-address registration-certificate --stake-verification-key-file stake.vkey --out-file stake.cert;"
Executed Command: /bin/bash -c "cardano-cli query utxo --mary-era --address $(cat payment.addr) --${CDN_NETCMD} ;"
Output Result: TxHash TxIx Amount
--------------------------------------------------------------------------------------
dd2d13c29f7e84ef4bd97f2dde7e412e79556687ad520beb0c1d1278a6100933 0 5000000 lovelace
Executed Command: /bin/bash -c "cardano-cli transaction build-raw --tx-in dd2d13c29f7e84ef4bd97f2dde7e412e79556687ad520beb0c1d1278a6100933#0 --tx-out $(cat payment.addr)+0 --invalid-hereafter 0 --fee 0 --out-file tx.draft --certificate-file stake.cert;"
Executed Command: /bin/bash -c "cardano-cli query protocol-parameters --mary-era --$CDN_NETCMD --out-file protocol.json;"
Executed Command: /bin/bash -c "cardano-cli transaction calculate-min-fee --tx-body-file tx.draft --tx-in-count 1 --tx-out-count 1 --witness-count 1 --protocol-params-file protocol.json;"
Output Result: 172453 Lovelace
Executed Command: /bin/bash -c "cardano-cli query protocol-parameters --mary-era --$CDN_NETCMD --out-file protocol.json;"
Executed Command: /bin/bash -c "awk '/keyDeposit/{print $2}' ${CDN_DIR}/protocol.json "
Output Result: 2000000,
Executed Command: /bin/bash -c "cardano-cli query tip --${CDN_NETCMD} | awk '/slotNo/{print $2}' "
Output Result: 19038051
Executed Command: /bin/bash -c "cardano-cli transaction build-raw --tx-in dd2d13c29f7e84ef4bd97f2dde7e412e79556687ad520beb0c1d1278a6100933#0 --tx-out $(cat payment.addr)+2827547 --invalid-hereafter 19038251 --fee 172453 --out-file tx.raw --certificate-file stake.cert;"
Executed Command: /bin/bash -c "cardano-cli transaction sign --tx-body-file tx.raw --signing-key-file stake.skey --signing-key-file payment.skey --out-file tx.signed;"
Executed Command: /bin/bash -c "cardano-cli transaction submit --tx-file tx.signed --$CDN_NETCMD ;"
```
-> Worked

```
Executed Command: /bin/bash -c "cardano-cli transaction build-raw --tx-in 92b27b99677713a0efd729613fd924d073bb4114fb6431bffd1fa7bda6bde293#0 --tx-out $(cat payment.addr)+655094 --invalid-hereafter 19038638 --fee 167453 --out-file tx.raw --certificate-file stake.cert;"
Executed Command: /bin/bash -c "cardano-cli transaction sign --tx-body-file tx.raw --signing-key-file stake.skey --signing-key-file payment.skey --out-file tx.signed;"
Executed Command: /bin/bash -c "cardano-cli transaction submit --tx-file tx.signed --$CDN_NETCMD ;"
Error Result: Shelley command failed: transaction submit Error: Error while submitting tx: ApplyTxError [LedgerFailure (DelegsFailure (DelplFailure (DelegFailure (StakeKeyAlreadyRegisteredDELEG (KeyHashObj (KeyHash "de92dd3db6c8471acdf4c6d49488a2b92983b6dccd6bd67ebfbd4cfb")))))),LedgerFailure (UtxowFailure (UtxoFailure (FeeTooSmallUTxO (Coin 171265) (Coin 167453)))),LedgerFailure (UtxowFailure (UtxoFailure (ValueNotConservedUTxO (Value 2827547 (fromList [])) (Value 2822547 (fromList []))))),LedgerFailure (UtxowFailure (UtxoFailure (OutputTooSmallUTxO [(Addr Testnet (KeyHashObj (KeyHash "20cdd18c69a22c80ccd97d6121b0eec36a6941093cf0080b643e947c")) (StakeRefBase (KeyHashObj (KeyHash "de92dd3db6c8471acdf4c6d49488a2b92983b6dccd6bd67ebfbd4cfb"))),Value 655094 (fromList []))])))]
```

Observation: The calucation gives me a minimum fee of *172453 Lovelace* while the Exception talks about *FeeTooSmallUTxO (Coin 171265)* so i assume that something is still wrong with the calulcation.

**Comment by hootan-nikbakht** [comment]
@tigrpoolcom calculated fee and the FeeTooSmallUTxO amount doesn't have to be the same. My calculated fee also came out much higher than FeeTooSmallUTxO amount since I used 3 for '--witness-count'. My calculated fee was 197269 Lovelace which is definitely much higher but I used 3 keys so I used 3 as '--witness-count'. 

Are you unblocked and able to register your pool ? if so, what is your ticker ?
